src/pool/srv_pool_scrub.c

```diff
@@ -4,324 +4,767 @@
 }
 
-static uint64_t
-between_scrub_sec()
+static void
+sc_obj_val_setup(struct scrub_ctx *ctx, vos_iter_entry_t *entry,
+		 vos_iter_type_t type, vos_iter_param_t *param,
+		 daos_handle_t ih)
 {
-	char *sec = getenv("DAOS_CSUM_SCRUB_PAUSE_SEC");
+	ctx->sc_cur_oid = param->ip_oid;
+	ctx->sc_dkey = param->ip_dkey;
+	ctx->sc_epoch = entry->ie_epoch;
 
-	return sec != NULL ? atoll(sec) : 24 * 60 * 60;
+	ctx->sc_iod.iod_size = entry->ie_rsize;
+	ctx->sc_iod.iod_nr = 1;
+	ctx->sc_iod.iod_type = type == VOS_ITER_RECX ? DAOS_IOD_ARRAY :
+			       DAOS_IOD_SINGLE;
+	ctx->sc_iod.iod_name = param->ip_akey;
+	ctx->sc_iod.iod_recxs = &entry->ie_recx;
+
+	ctx->sc_csum_to_verify = &entry->ie_csum;
+
+	ctx->sc_vos_iter_handle = ih;
+}
+
+static bool
+oids_are_same(daos_unit_oid_t a, daos_unit_oid_t b)
+{
+	return daos_unit_oid_compare(a, b) == 0;
+}
+
+static bool
+keys_are_same(daos_key_t key1, daos_key_t key2)
+{
+	if (key1.iov_len != key2.iov_len)
+		return 0;
+
+	return memcmp(key1.iov_buf, key2.iov_buf, key1.iov_len) == 0;
+}
+
+static bool
+epoch_is_same(daos_epoch_t a, daos_epoch_t b)
+{
+	return a == b;
 }
 
 /** vos_iter_cb_t */
 static int
-obj_iter_scrub_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+obj_iter_scrub_pre_cb(daos_handle_t ih, vos_iter_entry_t *entry,
 		  vos_iter_type_t type, vos_iter_param_t *param,
 		  void *cb_arg, unsigned int *acts)
 {
 	struct scrub_ctx	*ctx = cb_arg;
-	struct daos_csummer	*csummer = ctx->cur_cont->sc_csummer;
+	int			 rc;
 
-	if (!(type == VOS_ITER_RECX || type == VOS_ITER_SINGLE))
+	if (ctx->sc_pool->sp_scrub_sched == DAOS_SCRUB_SCHED_OFF) {
+		C_TRACE("scrubbing is off now, aborting ...");
+		*acts |= VOS_ITER_CB_ABORT;
 		return 0;
+	}
 
-	C_TRACE("Scrubbing akey: "DF_KEY", type: %s, record size: "
-			DF_U64", extent: "DF_RECX", csum_size: %d\n",
-		DP_KEY(&param->ip_akey),
-		(type == VOS_ITER_RECX) ? "ARRAY" : "SV",
-		entry->ie_rsize,
-		DP_RECX(entry->ie_orig_recx),
-		daos_csummer_get_csum_len(csummer)
-	);
-
-	/** TODO - implement the actual scrubbing of the checksum by fetching
-	 * the data, calculating a new checksum, and comparing to the stored
-	 * checksum.
-	 */
-	ctx->pool_csums_scrubbed++;
+	switch (type) {
+	case VOS_ITER_OBJ:
+		if (oids_are_same(ctx->sc_cur_oid, entry->ie_oid)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_cur_oid, 0, sizeof(ctx->sc_cur_oid));
+		} else {
+			ctx->sc_cur_oid = entry->ie_oid;
+		}
+		break;
+	case VOS_ITER_DKEY:
+		if (keys_are_same(ctx->sc_dkey, entry->ie_key)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_dkey, 0, sizeof(ctx->sc_dkey));
+		} else {
+			ctx->sc_dkey = param->ip_dkey;
+		}
+		break;
+	case VOS_ITER_AKEY:
+		if (keys_are_same(ctx->sc_iod.iod_name, entry->ie_key)) {
+			*acts |= VOS_ITER_CB_SKIP;
+			memset(&ctx->sc_iod, 0, sizeof(ctx->sc_iod));
+		} else {
+			ctx->sc_iod.iod_name = param->ip_akey;
+		}
+		break;
+	case VOS_ITER_SINGLE:
+	case VOS_ITER_RECX: {
+		if (epoch_is_same(ctx->sc_epoch, entry->ie_epoch)) {
```

Comment:
I don't think this is quite right.   You need to check minor epoch as well for distributed transaction case.   In a DTX, multiple VOS operations happen under same PMDK transaction and use the minor epoch as the differentiator.